### ISSUE TITLE
ProFTPd server flagged for "FTPS Cleartext Fallback Security Bypass"

### ISSUE THREAD

**Issue by ws0w** [Castaglia/proftpd-mod_proxy#192]
This is really more of a question that an issue report ....

Tenable has flagged two of my ProfTPD servers for a vulnerability:

FTPS Cleartext Fallback Security Bypass

The report says that after a failed TLS negotiation, the server responded to the "FEAT" command with a list of features the server supports.

I can telnet to port 21 and get a valid response to the FEAT command, but I'm not sure that's a vulnerability ... any other command gets the error: "550 SSL/TLS required on the control channel" (or "500 Invalid command: try being more creative")

I'm thinking that responding to FEAT in the clear is not really an example of the flagged vulnerability.

What are your thoughts on this?

**Comment by Castaglia** [comment]
IMHO, the `FEAT` command _should_ be accepted before any `AUTH TLS` command occurs; issuing a `FEAT` command is how a well-behaved client can "discover" that the server accepts/supports FTPS.

A brief reading of the [Tenable docs](https://www.tenable.com/plugins/nessus/57272) for this suggest that responding to `FEAT` -- by itself -- is not a problem.  Or rather, that the issue they are flagging is that a client tries do the TLS handshake first, and -- assuming that TLS handshake files -- then sends a `FEAT` request.  Their docs suggest that a "good" FTPS server, if it fails the TLS handshake, would close the control connection altogether.

I'll see if I can reproduce this behavior locally.  If that does indeed happen -- a good response to a `FEAT` command, sent after a failed TLS handshake _on the control connection_, then I would think it's a bug, in the `mod_tls` module.

**Comment by pedanticdan** [comment]
The report did claim that it was after a failed TLS handshake.

**Comment by Castaglia** [comment]
What ProFTPD version is running, on your flagged servers?  Reading through these docs:
* https://www.tenable.com/plugins/nessus/53847
* https://www.securityfocus.com/archive/1/516901/30/0/threaded

makes it sound exactly like [Bug #3624](http://bugs.proftpd.org/show_bug.cgi?id=3624), which _should have_ been fixed long ago.  (It's possible that it has regressed, hence the checking of the version...)

**Comment by Castaglia** [comment]
And, to muddy the waters further, the code seen [here](https://vulners.com/nessus/FTPS_PLAINTEXT_FALLBACK.NASL) (which purports to be that same Nessus plugin) uses port 990, not port 21.  ProFTPD doesn't use port 990 for FTPS connections unless configured to do so; are you using that configuration, for implicit FTPS connections?

**Comment by ws0w** [comment]
I get 1.3.7rc1 from proftpd -v .... we're using port 21 ...
I do have the TLS log from the time the scan was happening, but did not have TLSOptions EnableDiags enabled.

There are a lot of entries like this the day the scan happened, if it's useful:

2021-02-11 02:01:39,006 mod_tls/2.8[22872348]: TLS/TLS-C requested, starting TLS handshake
2021-02-11 02:01:39,007 mod_tls/2.8[22872348]: unable to accept TLS connection: protocol error:
2021-02-11 02:01:39,007 mod_tls/2.8[22872348]: unable to accept TLS connection: perhaps client requested disabled TLS protocol version: SSLv2, SSLv3, TLSv1, TLSv1.1
2021-02-11 02:01:39,007 mod_tls/2.8[22872348]: TLS/TLS-C negotiation failed on control channel

I was going to try to match it with FEAT commands using the pid, but the Deb 11 syslog is gone.

**Comment by ws0w** [comment]
I was able to retrieve the syslog for Feb 11 and I could not find a PID that issued FEAT commands and also had a TLS handshake failure in the TLS.log

My experience has been that Tenable's tests (especially related to FTP) frequently do not do what they think they do..

**Comment by ws0w** [comment]
It looks like a False positive to me,  but I could add EnableDiags and request  another scan if you think there's a change that would be helpful.

**Comment by Castaglia** [comment]
If it's not too much trouble, having them scan again would be useful, using _e.g_:
```
SystemLog /path/to/proftpd.log
DebugLevel 10

# Send mod_tls logs to the same file as normal ProFTPD logs
TLSLog /path/to/proftpd.log
TLSOptions EnableDiags
```
This way, we can see the FTP commands the client sends, along with the corresponding TLS commands/handshakes.

**Comment by Castaglia** [comment]
OK, no worries.  Thanks for the update!

**Comment by Castaglia** [comment]
I found that this ticket was still open, so I did a little more research/work on it, which results in some regression tests to assert that ProFTPD is indeed doing the right thing; see https://github.com/proftpd/proftpd/pull/1478.

I also ran the above tests against ProFTPD 1.3.7rc1 (and they passed); ProFTPD closes the FTP control connection if the TLS handshake fails, and does not leave that connection open in plaintext mode.

I'm thus going to close this ticket.  If we get similar reports from scanners/tools in the future, we'll take a look then.

Thanks!
